### PR TITLE
embedding OSBS config file

### PR DESCRIPTION
**Pull request by dbecvarik**

### CODE REVIEW

File: cekit/descriptor/osbs.py [hunk 1]
```diff
@@ -20,3 +28,27 @@ class Osbs(Descriptor):
...
+      descriptor - yaml contianing OSBS configuration"""
+
+    def __init__(self, descriptor):
+        self.schemas = configuration_schema
+        super(Configuration, self).__init__(descriptor)
+        self.skip_merging = ['container', 'container_file']
+
+    def _prepare_configuration(self):
+        if 'container' in self and 'container_file' in self:
+            raise CekitError('You cannot specifu container and container_file together!')
```
Comment (goldmann) [inline cekit/descriptor/osbs.py]: Typo.

File: cekit/descriptor/osbs.py [hunk 2]
```diff
@@ -20,3 +28,27 @@ class Osbs(Descriptor):
...
+
+        if 'container_file' in self:
+            with open(self['container_file'], 'r') as file_:
```
Comment (goldmann) [inline cekit/descriptor/osbs.py]: Don't we want some check if the file exists?

File: cekit/descriptor/packages.py [hunk 3]
```diff
@@ -69,28 +71,34 @@ def __init__(self, descriptor):
...
         if not (('url' in descriptor) ^
-                ('odcs' in descriptor) ^
                 ('id' in descriptor) ^
-                ('rpm' in descriptor)):
+                ('rpm' in descriptor) ^
+                ('content_sets' in descriptor) ^
+                ('content_sets_file' in descriptor)):
             raise CekitError("Repository '%s' is invalid, you can use only one of "
-                             "['id', 'odcs', 'rpm', 'url']"
+                             "['id', 'content_sets_file', 'content_sets', 'rpm', 'url']"
```
Comment (goldmann) [inline cekit/descriptor/packages.py]: I'm not sure if we want to remove odcs, this is a general type. content_sets is very specific to our environment.

File: cekit/descriptor/packages.py [hunk 4]
```diff
@@ -69,28 +71,34 @@ def __init__(self, descriptor):
...
-            descriptor['odcs'] = {'pulp': descriptor['id']}
-            del descriptor['id']
+        if ('content_sets_file' in descriptor and 'content_sets' in descriptor):
+            raise CekitError("You cannot specify content_sets and content_sets_file together!")
 
         if 'url' not in descriptor:
             descriptor['url'] = {}
 
+        if 'content_sets_file' in descriptor:
+            with open(descriptor['content_sets_file'], 'r') as file_:
```
Comment (goldmann) [inline cekit/descriptor/packages.py]: We read it twice? It is read in `osbs.py` too.

File: cekit/generator/docker.py [hunk 5]
```diff
@@ -15,18 +17,23 @@ def __init__(self, descriptor_path, target, builder, overrides, params):
...
 
-    def _prepare_repository_odcs_pulp(self, repo):
-        """Create pulp content set in ODCS and returns its url
+    def _prepare_content_sets(self, repo):
+        if not tools.cfg['common']['redhat']:
+            return False
+
+        arch = platform.machine()
+        if arch not in repo['content_sets']:
+            raise CekitError("There are not contet_sets defined for you platform '%s'!")
```
Comment (goldmann) [inline cekit/generator/docker.py]: Typo: your.

File: docs/descriptor/osbs.rst [hunk 6]
```diff
@@ -0,0 +1,57 @@
+Osbs
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: OSBS.

File: docs/descriptor/osbs.rst [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+----
+This section represents object we use to hint OSBS builder with a configuration which needs to be tweaked
+for successful and reproducible builds.
+
+It contains two main keys:
+
+Repository
+^^^^^^^^^^
+This keys serves as a hint which DistGit repository and its branch we use to push generated sources into.
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: Typo: this key.

File: docs/descriptor/osbs.rst [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        repository:
+              name: containers/redhat-openjdk-18
+              branch: jb-openjdk-1.8-openshift-rhel-7
+
+
+
+Configuration
+^^^^^^^^^^^^^
+This key is holding OSBS ``container.yaml`` file ( :ref:`docs<https://osbs.readthedocs.io/en/latest/users.html?highlight=container.yaml#image-configuration>`_ )
+``contianer.yaml`` file can be embedded in ``container_yaml`` key or inject from a file specified in ``container_yaml_file`` key.
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: Typo: container.yaml.

File: docs/descriptor/osbs.rst [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+Embedded
+""""""""
+In this case whole ``container.yaml`` file is embedded in an image descriptor.
+
+.. code:: yaml
+
+    osbs:
+        configuration:
+	  container:
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: Tabs?

File: docs/descriptor/packages.rst [hunk 10]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
+            - name: foo
+	      url:
+	        repository: https://web.example/foo.repo
+                gpg: https://web.exmaple/foo.gpg
+
+
+.. _repo_contentsets:
+
 
+ContentSets
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: Content sets.

File: docs/descriptor/packages.rst [hunk 11]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
+^^^^^^^^^^^
+ContentSets are tightly integrated to OSBS style of defining repositories in ``content_sets.yml`` file.
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: Content sets

File: docs/descriptor/packages.rst [hunk 12]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
 
-URL
-^^^^
-This approach enables you to download a yum repository file and corresponding GPG key. To do it, define
-repositories section in a way of:
+.. code:: yaml
+
+    packages:
+        repositories:
+            - name: all
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: `all`?

File: cekit/descriptor/packages.py (hunk 4 above)
Comment (dbecvarik) [reply cekit/descriptor/packages.py]: no its not.. in osbs.py container.yaml is read not contentsets.yml

File: cekit/descriptor/packages.py (hunk 3 above)
Comment (dbecvarik) [reply cekit/descriptor/packages.py]: We dont have any use case for it now. And it was never used at the end, so I will propose to not carry it over.

File: docs/descriptor/osbs.rst [hunk 13]
```diff
@@ -0,0 +1,57 @@
+OSBS
+----
+This section represents object we use to hint OSBS builder with a configuration which needs to be tweaked
+for successful and reproducible builds.
+
+It contains two main keys:
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: List them here, in a list.

File: docs/descriptor/osbs.rst [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+Repository
+^^^^^^^^^^
+This key serves as a hint which DistGit repository and its branch we use to push generated sources into.
+
+
+**Example:**
+.. code:: yaml
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: Wrong formatting, the code part is not rendered properly.

File: docs/descriptor/osbs.rst [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+.. code:: yaml
+
+    osbs:
+        configuration:
+            container_file: container.yaml
+
+
+and ``container.yaml`` file contains:
+.. code:: yaml
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: Wrong formatting, the code part is not rendered properly.

File: docs/descriptor/osbs.rst [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+Embedded
+""""""""
+In this case whole ``container.yaml`` file is embedded in an image descriptor.
+
+.. code:: yaml
+
+    osbs:
+        configuration:
+           container:
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: So, `container` or `container_yaml` should be used?

File: docs/descriptor/osbs.rst [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+Linked
+""""""
+
+In this case ``container.yaml`` file is save next to the image descriptor.
+
+.. code:: yaml
+
+    osbs:
+        configuration:
+            container_file: container.yaml
```
Comment (goldmann) [inline docs/descriptor/osbs.rst]: So, `container_file` or `container_yaml_file` as per doc above?

File: docs/descriptor/packages.rst [hunk 18]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
 .. code:: yaml
 
     packages:
         repositories:
-            - name: Extras
-              odcs:
-                  repository: rhel-7-extras-rpm
+            - name: foo
+	      url:
+	        repository: https://web.example/foo.repo
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: Tabs?

File: docs/descriptor/packages.rst [hunk 19]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
-^^^^
-This approach enables you to download a yum repository file and corresponding GPG key. To do it, define
-repositories section in a way of:
+.. code:: yaml
+
+    packages:
+        repositories:
+            - name: content_sets
+              content_sets:
+	        x86_64:
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: Tabs?

File: docs/descriptor/packages.rst [hunk 20]
```diff
@@ -67,35 +64,72 @@ image you should define repository in a following way:
...
 
-URL
-^^^^
-This approach enables you to download a yum repository file and corresponding GPG key. To do it, define
-repositories section in a way of:
+.. code:: yaml
+
+    packages:
+        repositories:
+            - name: content_sets
```
Comment (goldmann) [inline docs/descriptor/packages.rst]: Should we highlight name here? Is it required?

Comment (dbecvarik) [reply docs/descriptor/packages.rst]: no its not. it can be any name

File: cekit/generator/docker.py [hunk 21]
```diff
@@ -22,18 +23,23 @@ def __init__(self, descriptor_path, target, builder, overrides, params):
...
 
-    def _prepare_repository_odcs_pulp(self, repo):
-        """Create pulp content set in ODCS and returns its url
+    def _prepare_content_sets(self, content_sets):
+        if not config.cfg['common']['redhat']:
+            return False
+
+        arch = platform.machine()
+        if arch not in content_sets:
+            raise CekitError("There are not contet_sets defined for platform '%s'!")
```
Comment (goldmann) [inline cekit/generator/docker.py]: "There are no"

File: docs/descriptor/osbs.rst [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    osbs:
+        configuration:
+           container:
+              compose:
+                  pulp: True
+
+Linked
+""""""
+
+In this case ``container.yaml`` file is save next to the image descriptor.
```
[the remark this answers is not in the document]
Comment (goldmann) [reply docs/descriptor/osbs.rst]:
```suggestion
In this case ``container.yaml`` file is read from a file located next to the image descriptor.
```

Maybe like this?

File: docs/descriptor/packages.rst [hunk 23]
```diff
@@ -67,35 +64,69 @@ image you should define repository in a following way:
...
     packages:
-        repositories:
-            - name: foo
-	      url:
-	        repository: https://web.example/foo.repo
-                gpg: https://web.exmaple/foo.gpg
+        content_sets_file: content_sets.yml
+
+
+content_sets.yml located next to image descriptor:
```
Comment (goldmann) [inline docs/descriptor/packages.rst]:
```suggestion
``content_sets.yml`` located next to image descriptor:
```

File: cekit/descriptor/osbs.py (hunk 2 above)
Comment (dbecvarik) [reply cekit/descriptor/osbs.py]: check is there in code above now. So this should be ok

**Pull request merged**